File: apps/scraper/metadata.js
import yaml from "js-yaml";

/**
 * @typedef {Object} StoryMeta
 * @property {string} story_id
 * @property {string} title
 * @property {string} author
 * @property {string} [summary]
 * @property {string} [url]
 * @property {string[]} [tags]
 * @property {number} [word_count]
 * @property {number} [chapter_count]
 * @property {string} [language]
 */

/**
 * @param {StoryMeta} meta
 * @returns {string} YAML string for meta.yaml
 */
export function toYaml(meta) {
  const obj = {
    story_id: meta.story_id,
    title: meta.title,
    author: meta.author,
    summary: meta.summary ?? "",
    url: meta.url ?? "",
    tags: meta.tags ?? [],
    word_count: meta.word_count ?? 0,
    chapter_count: meta.chapter_count ?? 0,
    language: meta.language ?? "en",
  };
  return yaml.dump(obj, { lineWidth: -1, noRefs: true });
}

File: packages/scraper/parseltongue_scraper/metadata.py
from dataclasses import dataclass, field

import yaml


@dataclass
class StoryMeta:
    story_id: str
    title: str
    author: str
    summary: str = ""
    url: str = ""
    rating: list[str] = field(default_factory=list)
    category: list[str] = field(default_factory=list)
    tags: list[str] = field(default_factory=list)
    word_count: int = 0
    chapter_count: int = 0
    language: str = "en"


def to_yaml(meta: StoryMeta) -> str:
    obj = {
        "story_id": meta.story_id,
        "title": meta.title,
        "author": meta.author,
        "summary": meta.summary,
        "url": meta.url,
        "rating": meta.rating,
        "category": meta.category,
        "tags": meta.tags,
        "word_count": meta.word_count,
        "chapter_count": meta.chapter_count,
        "language": meta.language,
    }
    return yaml.dump(obj, default_flow_style=False, allow_unicode=True, width=-1)
